MSAL/test/integration/native_auth/end_to_end/mfa/MFADelegateSpies.swift

```diff
@@ -0,0 +1,131 @@
+//
+// Copyright (c) Microsoft Corporation.
+// All rights reserved.
+//
+// This code is licensed under the MIT License.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files(the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and / or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions :
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.  
+
+import Foundation
+import XCTest
+import MSAL
+
+class MFARequestChallengeDelegateSpy: MFARequestChallengeDelegate {
+    
+    private let expectation: XCTestExpectation
+    private(set) var onMFARequestChallengeError = false
+    private(set) var error: MSAL.MFAError?
+    
+    private(set) var onVerificationRequiredCalled = false
+    private(set) var newStateMFARequired: MSAL.MFARequiredState?
+    private(set) var sentTo: String?
+    private(set) var channelTargetType: MSAL.MSALNativeAuthChannelType?
+    private(set) var codeLength: Int?
+    
+    private(set) var onSelectionRequiredCalled = false
+    private(set) var authMethods: [MSALAuthMethod]?
+
+    init(expectation: XCTestExpectation) {
+        self.expectation = expectation
+    }
+    
+    func onMFARequestChallengeError(error: MSAL.MFAError, newState: MSAL.MFARequiredState?) {
+        onMFARequestChallengeError = true
+        self.newStateMFARequired = newState
+        self.error = error
+
+        expectation.fulfill()
```

Comment:
for my own understanding, what happens when this is never called? Does `XCTestExpectation` have a timeout?

Reply:
Expectations are used to test async scenario. In the test class, we declare the expectation, we pass in the delegate instance (this spy class), and we wait for expectation fulfilment specifying a timeout (example MSALNativeAuthSignInWithMFAEndToEndTests.swift line 47). Our default timeout is 20 seconds. If an exception isn't fulfilled in that time, the test fails